[oneDNN v3.x]: Enabled reorder primitive cache and some misc. changes

bhavani-subramanian opened the pull request:
- This PR enables reorder primitive cache in `mkl_util.h` and `mkl_quantize_op.cc`.
- This PR passes all unit tests related to reorder primitive cache for oneDNN v3.x.
- This PR also enables oneDNN v3.x in benchmark tests.
- These changes are specific only to oneDNN kernels and do not affect Eigen ops.

bhavani-subramanian committed:
[oneDNN v3.x]: Enabled reorder primitive cache and some misc. changes

bhavani-subramanian committed:
[oneDNN v3.x]: Enabled rewrite of int8 ops in mkl_layout_pass

bhavani-subramanian committed:
[oneDNN v3.x]: Added support for Eigen threadpool for conv bwd and softmax

bhavani-subramanian committed:
[oneDNN v3.x]: Fixed a bug in reorder primitive cache

bhavani-subramanian committed:
[oneDNN v3.x]: Fixed a bug in reorder primitive cache caused by non-unique keys

bhavani-subramanian committed:
Merge branch 'master' into bhavanis/onednn-3.0-misc

bhavani-subramanian committed:
Replaced all usage of errors::* with absl::*

penpornk reviewed (changes requested):
Thank you for the PR and sorry for the delay! I have one question.

bhavani-subramanian committed:
Fixed oneDNN v3.x build errors

bhavani-subramanian committed:
Merge branch 'master' into bhavanis/onednn-3.0-misc

bhavani-subramanian committed:
Merge branch 'master' into bhavanis/onednn-3.0-misc

bhavani-subramanian committed:
Fixed Clang build errors for oneDNN v3.x

bhavani-subramanian committed:
Wrapped reorder primitive cache within oneDNN v3.x macro

bhavani-subramanian commented on tensorflow/core/util/mkl_util.h:
```diff
@@ -2157,64 +2136,99 @@ class MklReorderPrimitiveFactory : public MklPrimitiveFactory<T> {
+    memory::dims to_inner_blks = to_desc.get_inner_blks();
+    memory::dims to_inner_idxs = to_desc.get_inner_idxs();
+    memory::dims to_strides = to_desc.get_strides();
+#endif  // !ENABLE_ONEDNN_V3
```
Makes sense. Fixed it.

bhavani-subramanian commented:
> Could you please merge conflicts and rebase? All tests are failing and, for some reason, there are no logs. Please also reply to my comment on `mkl_util.h`. Thank you!

@penpornk  Thanks for reviewing this PR! I have addressed your review comments and also rebased with latest master. Please take a look.

penpornk commented on tensorflow/core/kernels/mkl/mkl_fused_instance_norm_op.cc:
```diff
@@ -13,7 +13,7 @@ See the License for the specific language governing permissions and
 ==============================================================================*/
 
-#ifdef INTEL_MKL
+#if defined(INTEL_MKL) && !defined(ENABLE_ONEDNN_V3)
```
Could you please add `&& !defined(ENABLE_ONEDNN_V3)` in the comment at the `#endif` line at the end of the file as well?

bhavani-subramanian replied:
@penpornk Looks like this kernel got disabled for oneDNN v3.x after rebasing with master. I have fixed it now. Please take a look.

penpornk replied:
Got it. Thank you!

penpornk reviewed (changes requested):
Thank you for the changes! One minor nit please.

bhavani-subramanian committed:
Re-enabled instance-norm kernel for oneDNN v3.x

bhavani-subramanian committed:
Merge branch 'master' into bhavanis/onednn-3.0-misc

penpornk reviewed (approved):
Thank you for the sync!
